lib/charms/layer/kubernetes_master.py

```diff
@@ -309,6 +315,32 @@ def get_csv_password(csv_fname, user):
     return None
 
 
+def get_secret_names():
+    '''Return a dict of 'username: secret_id' for Charmed Kubernetes users.'''
+    try:
+        output = kubernetes_common.kubectl(
+            'get', 'secrets', '-n', AUTH_SECRET_NS,
+            '--field-selector', 'type={}'.format(AUTH_SECRET_TYPE),
+            '-o', 'json').decode('UTF-8')
+    except (CalledProcessError, FileNotFoundError):
+        # The api server may not be up, or we may be trying to run kubelet before
+        # the snap is installed. Send back an empty dict.
+        return {}
```

Comment:
I would at least log a warning here

Reply:
If the kubectl call here fails and we return `{}`, then create_secret will move on as if the secret doesn't exist, and try to create a new secret. We could end up with duplicate secrets in that case right?

Reply:
> If the kubectl call here fails and we return `{}`, then create_secret will move on as if the secret doesn't exist, and try to create a new secret. We could end up with duplicate secrets in that case right?

That's true, but if the apiserver is down for `get_secret_names`, then it will *probably* still be down for the subsequent `kubectl apply`.  I plan to refactor the handling of a "down" apiserver in 1.21.